File: leetcode-new/22.cpp
class Solution
{
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> res;
        helper(res, n, 0, 0, "");
        return res;
    }
    void helper(auto &res, int n, int o, int c, string cand)
    {
        if (cand.length() == n * 2)
        {
            res.push_back(cand);
        }
        if (o < n)
        {
            helper(res, n, o + 1, c, cand + "(");
        }
        if (c < o)
        {
            helper(res, n, o, c + 1, cand + ")");
        }
    }
};

File: leetcode-venture/22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        res = []

        def backtrack(usedL, usedR, cand):
            if usedL == usedR == n:
                res.append(cand)
                return
            if usedL < n:
                backtrack(usedL+1, usedR, cand+"(")
            if usedR < usedL:
                backtrack(usedL, usedR+1, cand+")")

        backtrack(0,0,"")
        return res
